### PR TITLE
Added numeric field with isNumeric constraint.

### PR DESCRIPTION
**Pull request by Amstutz**
This afternoon I finally had the opportunity to play with the proposed inputs. To start out easy, I implemented I new input type "numeric" with the according isNumeric constraint. Some questions popped up during that (especially concerning the private nature of constraints in input). In order to proceed in small manegable chunks for discussion I created this rather smallish PR along with some questions (look out for "TODOs").

Looking forward to proceed with this project.

### CODE REVIEW

File: src/UI/Implementation/Component/Input/Input.php [hunk 1]
```diff
@@ -174,6 +184,7 @@ public function withError($error) {
 	// input is processed.
 
 	/**
+	 * TODO: Should this not be "withAdditionalTransformation"?
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Input.php]: 👍

File: src/UI/Implementation/Component/Input/Input.php [hunk 2]
```diff
@@ -189,6 +200,7 @@ public function withTransformation(Transformation $trafo) {
 	}
 
 	/**
+	 * TODO: Should this not be "withAdditionalConstraint"?
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Input.php]: 👍

File: src/UI/Implementation/Component/Input/Input.php [hunk 3]
```diff
@@ -60,9 +60,19 @@
...
 	/**
+	 * TODO: There must be a way to access those in order to one day get the constraints
+	 * for possible client side checks. Also, I guess the required attribute would also
+	 * be handled by passing a "NotNull" constraint, correct? If yes the renderer needs
+	 * to be able to check if such a constraint is passed for possible special labeling
+	 * of required fields.
+	 * Also, note that the operations can not be attached in the constructor due to
+	 * clone in method withConstraints.
+	 *
+	 * ==> currently changed to protected
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Input.php:71]: I would rather provide some protected setters to protect these internal mechanisms from tempering.

File: src/UI/Implementation/Component/Input/Input.php [hunk 4]
```diff
@@ -60,9 +60,19 @@
 	private $content;
 
 	/**
+	 * TODO: There must be a way to access those in order to one day get the constraints
+	 * for possible client side checks. Also, I guess the required attribute would also
+	 * be handled by passing a "NotNull" constraint, correct? If yes the renderer needs
+	 * to be able to check if such a constraint is passed for possible special labeling
+	 * of required fields.
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Input.php:67]: I agree with the getter for client side checks. I'm not sure though, if `NotNull` is the correct implementation for required. Imo this will interfere badly with `withAdditionalTransformation`.

File: src/UI/Component/Form/Form.php [hunk 5]
```diff
@@ -44,5 +44,9 @@ public function withTransformation(Transformation $trafo);
 	 */
 	public function getData();
 
+	/**
+	 * TODO: there should be a further method to attach the different submit buttons
+	 */
```
Comment (klees) [inline src/UI/Component/Form/Form.php:49]: What kind of buttons would that be?

File: tests/Validation/Constraints/IntConstraintsTest.php [hunk 6]
```diff
@@ -1,87 +0,0 @@
...
-
-use ILIAS\Validation;
-use ILIAS\Data;
-
-/**
- * TestCase for the factory of constraints
- *
- * @author Stefan Hecken <stefan.hecken@concepts-and-training.de>
- */
-class IntConstraintsTest extends PHPUnit_Framework_TestCase {
```
Comment (klees) [inline tests/Validation/Constraints/IntConstraintsTest.php:13]: Why did you remove this?

**Review by klees — COMMENTED** [review]
Hi Timon,
added some comments. Please propose a date for a skype meeting for discussion. I wonder where the Renderer of the new input is, though. Am I blind? Is it missing?
Regards

File: tests/Validation/Constraints/IntConstraintsTest.php (hunk 6 above)
Comment (Amstutz) [reply tests/Validation/Constraints/IntConstraintsTest.php:13]: Because almost all of this code can be reused for non-int tests as well (see isNumeric). Just renamed it to standard and added just a slight bit of more flexibility (to process mulitple ok's and errors at once).

Comment (klees) [reply tests/Validation/Constraints/IntConstraintsTest.php:13]: Ah, thx!

File: src/UI/Component/Form/Form.php (hunk 5 above)
Comment (Amstutz) [reply src/UI/Component/Form/Form.php:49]: Different forms might need different buttons to perform actions with/handle the form. E.g. "Save, Send, Cancel, Next, Back" ...

It is possible but we can do with just changing the labels with some function like "withSubmitLabel" with "Save" as default. But we might still need additional regular buttons such as "Cancel".

**Comment by Amstutz** [comment]
> I wonder where the Renderer of the new input is, though. Am I blind? Is it missing?

None of those. I was lazy. I did not care about rendering for the first part. The rendering only worked by accident, since my implementation implemented the wrong (but identical) interface.

I just added a proposal for a rendering and stumbled again over the context question. When implementing inputs for various containers (such as a form) they will most certainly wrapped by a different dom. I factored this out on a template level. However for this to work smoothly we need the context proposed in: https://github.com/ILIAS-eLearning/ILIAS/pull/532
